### PR TITLE
[Fiber] Move memoization to begin phase

### PR DESCRIPTION
**Pull request by acdlite**
Because `memoizedProps` and `memoizedState` are read from the instance, the
instance's input pointers (`props`, `state`, `context`) should be updated even
when `shouldComponentUpdate` causes a bail-out.

## Update

The scope of this PR changed a bit. The main change is to move memoization to the begin phase, right after reconciliation.

Currently we update the memoized inputs (props, state) during the complete phase, as we go back up the tree. That means we can't reuse work until of its children have completed.

By moving memoization to the begin phase, we can do a shallow bailout, reusing a unit of work even if there's still work to do in its children.

Memoization now happens whenever a fiber's `child` property is updated; typically, right after reconciling. It's also updated when `shouldComponentUpdate` returns false, because that indicates that the given state and props are equal to the memoized state and props.

### CODE REVIEW

**Comment by acdlite** [comment]
If you can think of a more realistic test scenario let me know. The one I wrote is pretty contrived, but since this is an edge case that might be unavoidable.

**Comment by acdlite** [comment]
I'll write one and see if it can be solved by setting the priority level at the same time as memoization, as you suggested in the FB group.

**Comment by acdlite** [comment]
@sebmarkbage Struggling to come up with a failing test case. As far as I can tell, even when work is aborted half way, the subtree doesn't get reused because `bailoutOnFinishedWork` returns the child.

**Comment by sebmarkbage** [comment]
Try it with one or two more levels. The work of the direct child isn't reused because one more child needs to be worked on to test if it has priority level. Also, ensure that there are siblings since currently work can only be reused after a tree is fully completed so to abort you need to be able to abort in a later sibling.

File: src/renderers/shared/fiber/ReactFiberClassComponent.js [hunk 1]
```diff
@@ -103,6 +103,8 @@ module.exports = function(
 
     const instance = workInProgress.stateNode;
     if (typeof instance.shouldComponentUpdate === 'function') {
+      instance.props = workInProgress.memoizedProps;
+      instance.state = workInProgress.memoizedState;
       const shouldUpdate = instance.shouldComponentUpdate(newProps, newState, newContext);
```
Comment (bvaughn) [inline src/renderers/shared/fiber/ReactFiberClassComponent.js]: Don't we have the same potential problem below for the `ReactPureComponent` state check?
```jsx
!shallowEqual(instance.state, newState)
```

Shouldn't this function always use `memoizedProps` and `memoizedState`? (Can we also just remove the redundant `oldProps` parameter?)

File: src/renderers/shared/fiber/ReactFiberBeginWork.js [hunk 2]
```diff
@@ -581,20 +593,36 @@ module.exports = function<T, P, I, TI, C, CX>(
...
+        const next = updateFunctionalComponent(current, workInProgress);
+        memoizeProps(workInProgress);
+        return next;
+      }
+      case ClassComponent: {
+        const next = updateClassComponent(current, workInProgress, priorityLevel);
+        const instance = workInProgress.stateNode;
+        workInProgress.memoizedState = instance.state;
+        workInProgress.memoizedProps = instance.props;
+        workInProgress.pendingProps = null;
```
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: We had trouble keeping things in sync when we split these. It might be helpful to move this inside of `updateClassComponent` and similarly for the others so that it is easier to tell when this needs to happen relative to other things. I'm not 100% sure this should definitely be at the very end in all cases.

File: src/renderers/shared/fiber/ReactFiberClassComponent.js (hunk 1 above)
Comment (bvaughn) [inline src/renderers/shared/fiber/ReactFiberClassComponent.js]: Shouldn't we also do this reset before `componentWillReceiveProps`? and `componentWillUpdate`?

I think this argues for putting the reset at the start of `updateClassInstance`- so the other 2 lifecycle methods aren't called with stale props. Also, this way, even if the user doesn't define `shouldComponentUpdate` the instance attributes will still be synced before `render` is called.

**Comment by acdlite** [comment]
Found a few more problems. The test I wrote is wrong. I think relying so much on `instance.props` and `instance.state` is masking issues that we didn't notice before. Will look at it some more tomorrow.

**Comment by acdlite** [comment]
Ok, I clean up my commits and this is now ready for review.

I couldn't find a clean way to separate updating the instance pointers from moving memoization to the begin phase, so I've left them as part of the same commit. The diff isn't that big so hopefully it's not hard to review.

@sebmarkbage, I wrote a test to confirm that we don't drop work in the subtree. I tried to break this for a while and I'm pretty confident that it works.

File: src/renderers/shared/fiber/__tests__/ReactIncremental-test.js [hunk 3]
```diff
@@ -541,6 +541,80 @@ describe('ReactIncremental', () => {
...
+
+    function Parent() {
+      ops.push('Parent');
+      return [
+        // The extra div is necessary because when Parent bails out during the
+        // high priority update, its progressedPriority is set to high.
+        // So its direct children cannot be reused when we resume at
+        // low priority. I think this would be fixed by changing
+        // pendingWorkPriority and progressedPriority to be the priority of
+        // the children only, not including the fiber itself.
```
Comment (acdlite) [inline src/renderers/shared/fiber/__tests__/ReactIncremental-test.js]: I started working on a fix for this, which I'll submit as a separate PR.

File: src/renderers/shared/fiber/ReactFiberBeginWork.js [hunk 4]
```diff
@@ -683,6 +712,14 @@ module.exports = function<T, P, I, TI, PI, C, CX>(
     // Unmount the current children as if the component rendered null
     const nextChildren = null;
     reconcileChildren(current, workInProgress, nextChildren);
+
+    if (workInProgress.tag === ClassComponent) {
+      const instance = workInProgress.stateNode;
+      workInProgress.memoizedProps = instance.props;
+      workInProgress.memoizedState = instance.state;
```
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: Why is this needed? Wouldn't this already have been memoized if a child threw?

Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: I put the memoization after reconciling the children because it seemed to make the most sense conceptually. Perhaps I could move it earlier before any lifecycles. The tricky part about that is that some things need to read from the previous memoized values. So you'd have to pass those around and it gets a bit hard to follow. But there's likely a better way to structure all this.

Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: I'd like to get to the point where we never read from the instance at all. This is one of two (I believe) places where we still do.

**Comment by sebmarkbage** [comment]
The heavy use of abstractions like markUpdate, markUpdateIfAlreadyInProgress, resetInputPointers, memoizeProps, memoizeState is an indication that we probably need to reorganize some of this so that those steps can be consolidated but we can do that later if this works out.

File: src/renderers/shared/fiber/ReactFiberBeginWork.js [hunk 5]
```diff
@@ -91,7 +90,12 @@ module.exports = function<T, P, I, TI, PI, C, CX>(
     mountClassInstance,
     resumeMountClassInstance,
     updateClassInstance,
-  } = ReactFiberClassComponent(scheduleUpdate, getPriorityContext);
+  } = ReactFiberClassComponent(
+    scheduleUpdate,
+    getPriorityContext,
+    memoizeProps,
```
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: Is this an abstraction worthy the pain of passing around? :)

Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: Haha I almost did it so I could log when I was happening :D And at one point it was a bit more involved.

I agree this could be restructured in a better way but when I'm unsure I prefer to just find something that works and passes the tests and then clean it up later.

Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: *mostly, not almost

Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: For example if I rewrote all the stuff in ReactFiberClassComponent, and then you told me the overall approach was wrong, I'd have wasted a bunch of time. And it would be harder to review. Easier to rewrite after overall approach is figured out.

File: src/renderers/shared/fiber/ReactFiberBeginWork.js (hunk 4 above)
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: I still don't understand why this needs to be done in `beginFailedWork` specifically.

If anything, this should set the memoized values to `null` since this component is now in an invalid state. The children output doesn't correspond to the input.

Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: I agree but that breaks an assumption we make in a few places that we always have either pending or memoized props. I figured that was outside the scope of this PR but I'll fix it here if it helps this land. Shouldn't be too difficult.

Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: Let's put this way. Why are these lines in this PR at all? What happens if you just exclude them like it was before?

Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: A bunch of tests fail because of this invariant: https://github.com/facebook/react/blob/24b8ba7340f5ce66df0b858ead480ca7103a8153/src/renderers/shared/fiber/ReactFiberClassComponent.js#L329

Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: That wasn't the case before because we would always memoize in `completeWork`, even if it failed in `beginWork`.

Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: Haha if I remove those invariants everything seems to still work. I guess I thought it'd be a bit more complicated than that. I'll go ahead and push that :)

Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: Hmm although what should get passed to `shouldComponentUpdate` the next time if we set `memoizedProps` to `null`? An empty object?

Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: I guess in that case we shouldn't call `shouldComponentUpdate` at all since it's as if it's mounting for the first time.

Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: I think that something is broken here with this error handling logic but I understand now that you're only preserving existing behavior by moving it in here. I think that's fine for the first iteration.

**Comment by acdlite** [comment]
@sebmarkbage My understanding is: right now `bailoutOnAlreadyFinishedWork` returns the child, so it's not possible to lose the side-effects in the subtree. If we want to return null instead and do a full bailout, we need to ensure that the subtree has completed, so we don't lose side-effects. For that will need to flip a bit in the effect tag (or equivalent solution) as you pointed out.

Related: https://github.com/facebook/react/pull/8716

**Comment by acdlite** [comment]
Once the CI passes I'll merge this then open a follow-up PR to restructure and clean-up the weird abstractions.
